File: src/com/minhle/practiceleetcode/MinimumFallingPathSum.java
package com.minhle.practiceleetcode;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

public class MinimumFallingPathSum {
    int[][] dp;
    public int minFallingPathSum(int[][] matrix) {
        int n = matrix.length;
        dp = new int[n][n];
        System.arraycopy(matrix[n - 1], 0, dp[n - 1], 0, n);
        for (int r = n - 2; r >= 0; r--) {
            for (int c = 0; c < n; c++) {
                int x = Integer.MAX_VALUE;
                int y = Integer.MAX_VALUE;
                int z = Integer.MAX_VALUE;
                try {
                    x = dp[r + 1][c - 1];
                }
                catch (Exception ignored) {

                }
                try {
                    y = dp[r + 1][c];
                }
                catch (Exception ignored) {

                }
                try {
                    z = dp[r + 1][c + 1];
                }
                catch (Exception ignored) {

                }
                dp[r][c] = matrix[r][c] + Math.min(x, Math.min(y, z));
            }
        }
        int res = Integer.MAX_VALUE;
        for (int i = 0; i < n; i++) {
            res = Math.min(res, dp[0][i]);
        }
        return res;
    }


    @Test
    void test1() {
        int[][] matrix = {
                {-19, 57},
                {-40, -5}
        };
        Assertions.assertEquals(-59, minFallingPathSum(matrix));
    }

    @Test
    void test2() {
        int[][] matrix = {
                {2,1,3},
                {6,5,4},
                {7,8,9}
        };
        Assertions.assertEquals(13, minFallingPathSum(matrix));
    }
}

File: src/com/minhle/practiceleetcode/MinimumFallingPathSum.cpp
#include "AllNecessaryHeaders.h"

class Solution {
public:
    int minFallingPathSum(vector<vector<int>>& matrix) {
        int m = matrix.size();
        int n = matrix[0].size();
        vector<vector<int>> dp(m, vector<int>(n, INT_MIN));
        int res = INT_MAX;
        for (int c = 0; c < matrix[0].size(); ++c)
        {
            res = min(res, recursion(matrix, 0, c, dp));
        }
        return res;
    }

    int recursion(const vector<vector<int>> &matrix, int r, int c, vector<vector<int>> &dp)
    {
        if (c < 0 || c == matrix[r].size())
        {
            return INT_MAX;
        }
        if (r == matrix.size() - 1)
        {
            dp[r][c] = matrix[r][c];
            return dp[r][c];
        }
        if (dp[r][c] != INT_MIN)
        {
            return dp[r][c];
        }
        int min_sum = INT_MAX;
        for (int i = -1; i <= 1; ++i)
        {
            min_sum = min(min_sum, recursion(matrix, r + 1, c + i, dp));
        }
        dp[r][c] = matrix[r][c] + min_sum;
        return dp[r][c];
    }
};